flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -171,7 +171,20 @@ private static void visitFactSide(
                     joinKeys.stream()
                             .map(i -> scan.getRowType().getFieldNames().get(i))
                             .collect(Collectors.toList());
-            factSideFactors.isSuitableFactScanSource = !candidateFields.isEmpty();
+            if (candidateFields.isEmpty()) {
+                factSideFactors.isSuitableFactScanSource = false;
+                return;
+            }
+
+            List<String> suitableFields = new ArrayList<>();
+
+            for (String candidateField : candidateFields) {
```

Comment:
please add some comments to explain why we need this change

Reply:
Done!

Reply:
That is not the real reason. The reason is many unexpected query pattern will be treated as dpp if the candidateFields are not partition keys.